### PR TITLE
feat(dashboards): add all filter options + value suggestions to widget and monitor forms (LFE-9570)

### PR DESCRIPTION
**Pull request by marksalpeter**
🟡 **Live preview: [pr-15166.preview.langfuse.com](https://pr-15166.preview.langfuse.com)**

## Summary

The dashboard widget and monitor filter builders were missing most filter columns and offered no value suggestions, making dashboard filtering nearly unusable.

To achieve this we:
- Added the missing filter columns — observation/trace names, user, session, metadata keys & values, version, releases, experiment name/dataset/id, trace release/version, and score name/value — to the widget and monitor filter builders, populated from the filter-options APIs per view.
- Added a metadata key/value suggestions endpoint and hook so metadata filters autocomplete keys and their values within the selected time range.
- Consolidated the widget and monitor filter UIs into one shared `MetricsFilterBuilder` (folding the V1/V2 and per-view logic together) and extracted widget import into a `WidgetImporter` component.
- Switched metadata key/value entry to a creatable single-select so custom keys and values can be typed or picked.
- Redesigned the inline filter builder as a compact two-line layout (field/key on top, operator and value below with an elbow connector) so it stays legible in the narrow form panels, in light and dark mode.

**Note**: The compact two-line redesign applies to the inline filter builder used in forms; the table/dashboard popover filter keeps its existing single-row layout.

Fixes [LFE-9570](https://linear.app/langfuse/issue/LFE-9570/add-all-filter-options-to-dashboards-currently-unusable)

## Verification

- [x] `pnpm --filter web run lint`
- [x] `pnpm --filter web run typecheck`
- [x] `pnpm --filter web run test-client` — filter-builder, MetricsFilterBuilder, MonitorForm, WidgetImporter, useMetadataValueOptions (31 tests)
- [ ] `pnpm --filter @langfuse/shared run test event-query-builder` (CI)
- [x] Browser review of the widget and monitor forms, light and dark mode

## Test plan

- [ ] Create a widget: open Filters and confirm all columns appear for the selected view (metadata, user, session, names, versions, releases, experiments, scores).
- [ ] Pick a metadata filter: confirm key suggestions load, then value suggestions load for the chosen key; the operator and value stay disabled until a key is selected.
- [ ] Select a multi-value column: confirm the value field shows the selected chips (or "N selected") rather than a "Select" placeholder.
- [ ] Repeat in the monitor creation form and confirm parity with the widget form.
- [ ] Import a widget via the Import button and confirm it still works after the WidgetImporter extraction.
- [ ] Switch views (observations / scores-numeric / scores-categorical) and confirm the available filter columns update accordingly.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR significantly improves the dashboard widget and monitor filter builders by adding missing filter columns (user, session, metadata, version, release, experiment, score name/value) and value suggestions, consolidating the V1/V2 filter logic into a shared `MetricsFilterBuilder`, and introducing a compact two-line layout for the inline filter builder in narrow form panels.

- Adds two new ClickHouse query functions (`buildEventsMetadataKeysQuery` / `buildEventsMetadataValuesQuery`) and exposes them as new TRPC procedures (`events.metadataKeys` / `events.metadataValues`); the `useMetadataValueOptions` hook wires these queries to the filter builder so per-key value suggestions load on demand.
- Introduces a `labeledScalar` kind to the filter-option query builder so `experimentId` options carry a human-readable `displayValue` (experiment name) from the same ClickHouse scan; extends the result tuple from 4 to 5 elements and updates all callers.
- Extracts the widget import flow into a standalone `WidgetImporter` component and drops the now-redundant translation step from `WidgetForm`, since `MetricsFilterBuilder` owns that translation internally.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

Safe to merge. The new ClickHouse query paths are parameterized, authorization follows established patterns, and the filter UI changes are additive.

The core logic — labeledScalar tuple expansion, metadata key/value query builders, MetricsFilterBuilder consolidation — is correct and backed by a solid set of unit tests. Two minor issues exist: a style concern around the implicit `{metadataKey: String}` parameter sharing in `buildEventsMetadataValuesQuery`, and a UX edge case where `SingleSelect` propagates an empty string on deselect rather than a no-op. Neither affects data correctness or security.

`packages/shared/src/server/queries/clickhouse-sql/event-filter-options.ts` (implicit parameter binding coupling) and `web/src/features/filters/components/single-select.tsx` (deselect emits empty string).
</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant U as User (Widget/Monitor Form)
    participant MFB as MetricsFilterBuilder
    participant IFB as InlineFilterBuilder
    participant TK as events.metadataKeys (TRPC)
    participant TV as events.metadataValues (TRPC)
    participant FO as events.filterOptions (TRPC)
    participant CH as ClickHouse

    U->>MFB: open filter panel (projectId, view, dateRange)
    MFB->>FO: query(projectId, startTimeFilter)
    FO->>CH: buildEventsFilterOptionsForColumnsQuery (top-k bulk)
    CH-->>FO: filterOptions (incl. displayValue for experimentId)
    FO-->>MFB: filterOptions
    MFB->>TK: query(projectId, startTimeFilter)
    TK->>CH: buildEventsMetadataKeysQuery
    CH-->>TK: "[{value: key, count: N}, ...]"
    TK-->>MFB: metadataKeys
    MFB->>IFB: columns, stringObjectValueOptions, filters
    U->>IFB: select metadata key
    IFB->>MFB: onStringObjectKeyChange(key)
    MFB->>TV: query(projectId, key, startTimeFilter)
    TV->>CH: buildEventsMetadataValuesQuery
    CH-->>TV: "[{value: v, count: N}, ...]"
    TV-->>MFB: metadataValueOptions[key]
    MFB->>IFB: updated stringObjectValueOptions
    U->>IFB: select value
    IFB-->>MFB: onChange(FilterState)
    MFB-->>U: editorFiltersToViewFilters to canonical FilterState
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant U as User (Widget/Monitor Form)
    participant MFB as MetricsFilterBuilder
    participant IFB as InlineFilterBuilder
    participant TK as events.metadataKeys (TRPC)
    participant TV as events.metadataValues (TRPC)
    participant FO as events.filterOptions (TRPC)
    participant CH as ClickHouse

    U->>MFB: open filter panel (projectId, view, dateRange)
    MFB->>FO: query(projectId, startTimeFilter)
    FO->>CH: buildEventsFilterOptionsForColumnsQuery (top-k bulk)
    CH-->>FO: filterOptions (incl. displayValue for experimentId)
    FO-->>MFB: filterOptions
    MFB->>TK: query(projectId, startTimeFilter)
    TK->>CH: buildEventsMetadataKeysQuery
    CH-->>TK: "[{value: key, count: N}, ...]"
    TK-->>MFB: metadataKeys
    MFB->>IFB: columns, stringObjectValueOptions, filters
    U->>IFB: select metadata key
    IFB->>MFB: onStringObjectKeyChange(key)
    MFB->>TV: query(projectId, key, startTimeFilter)
    TV->>CH: buildEventsMetadataValuesQuery
    CH-->>TV: "[{value: v, count: N}, ...]"
    TV-->>MFB: metadataValueOptions[key]
    MFB->>IFB: updated stringObjectValueOptions
    U->>IFB: select value
    IFB-->>MFB: onChange(FilterState)
    MFB-->>U: editorFiltersToViewFilters to canonical FilterState
```

</a>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
packages/shared/src/server/queries/clickhouse-sql/event-filter-options.ts:496-510
**`{metadataKey: String}` binding provided only in one `whereRaw` call, but referenced in three query locations**

`valueAccessor` embeds `{metadataKey: String}` and is used in (a) the `selectExpression` passed to the constructor, (b) the second `.whereRaw(...)` that supplies no params object, and (c) the first `.whereRaw(...)` that supplies `{ metadataKey: params.key }`. This works because `buildWithParams()` merges all per-call params into a single map that ClickHouse receives, so one binding satisfies every occurrence. However, the binding for the uses in (a) and (b) is invisible at their call sites; if the builder's param-accumulation behaviour ever changes, or if someone refactors the two `whereRaw` calls independently, the query will fail at runtime with an unbound-parameter error. Adding an explicit comment noting this dependency, or providing the binding at the constructor level alongside the `selectExpression`, would make the contract clearer.

### Issue 2 of 2
web/src/features/filters/components/single-select.tsx:126-128
**Re-clicking a selected item commits `""` rather than a no-op or `undefined`**

`commit(isSelected ? "" : option.value)` will call `onValueChange("")` when the user clicks the currently selected option, which is the filter builder's metadata key. The parent `FilterBuilderForm` receives `{ ...filter, key: "" }` — a `stringObject` filter row with an empty key. In compact mode `keyPending` correctly greys out the operator/value controls, but the row stays in the filter state as a dangling, unmatchable condition. Users who expect "click the selected item to undo" get a different result: the filter row persists and looks broken until they click the row's delete button. Consider keeping the deselect gesture as a no-op (i.e. only commit when `!isSelected`) for the key picker context, or clear the whole row explicitly.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["feat(filters): compact two-line layout f..."](https://github.com/langfuse/langfuse/commit/926589c5b1145cedacffb98ca86fc97a73bb4a72) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44818286)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: web/src/features/widgets/components/WidgetForm.tsx [hunk 1]
```diff
@@ -476,17 +442,13 @@ export function WidgetForm({
...
   );
-  const unsupportedFilters = useMemo(
-    () =>
-      partitionWidgetUiTableFiltersToView(selectedView, userFilterState)
-        .unsupportedFilters,
+  const { mappedFilters: normalizedUserFilters, unsupportedFilters } = useMemo(
+    () => partitionWidgetUiTableFiltersToView(selectedView, userFilterState),
     [selectedView, userFilterState],
   );
   const unsupportedFilterColumns = useMemo(
```
[the remark this answers is not in the document]
Comment (marksalpeter) [reply web/src/features/widgets/components/WidgetForm.tsx:384]: Thats intended. We'd like to keep the names the same across views.

File: web/src/features/filters/components/single-select.tsx [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+      <PopoverContent className="w-[200px] p-0" align="center">
+        <InputCommand shouldFilter={false}>
+          <InputCommandInput
+            placeholder={title}
+            variant="bottom"
+            value={search}
+            onValueChange={setSearch}
+          />
+          <InputCommandList>
+            {filteredOptions.length === 0 && !showCustomValue && (
```
[the remark this answers is not in the document]
Comment (marksalpeter) [reply web/src/features/filters/components/single-select.tsx:129]: Fixed in 5c733dcb3: re-clicking a selected option is now inert — it closes the popover without committing. Confirmed worse than described: `stringObjectFilter.key`/`value` are `z.string()`, so `key: ""` passes `singleFilter.safeParse` and the row is pushed into the *applied* filter state, not just left dangling in WIP. Fixing it in `SingleSelect` covers both the key and value pickers.

File: web/src/features/filters/components/single-select.tsx [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                    value={option.value}
+                    onSelect={() => commit(isSelected ? "" : option.value)}
+                  >
+                    <Check
+                      className={cn(
+                        "mr-2 h-4 w-4",
+                        isSelected ? "visible" : "invisible",
+                      )}
+                    />
+                    <div
```
[the remark this answers is not in the document]
Comment (marksalpeter) [reply web/src/features/filters/components/single-select.tsx]: resolved https://github.com/langfuse/langfuse/pull/15166#discussion_r3757979262

File: web/src/features/monitors/components/MonitorForm.tsx [hunk 4]
```diff
@@ -614,11 +487,13 @@ export const MonitorForm = ({
...
+                          version="v2"
+                          view={(watched.view ?? "observations") as MonitorView}
+                          projectId={projectId}
+                          dateRange={{ from: filterOptionsLookbackFrom }}
+                          filters={(field.value ?? []) as FilterState}
                           onChange={(next: FilterState) => field.onChange(next)}
-                          columnsWithCustomSelect={customSelectColumnIds}
                         />
                       </FormControl>
                       <FormMessage />
```
[the remark this answers is not in the document]
Comment (marksalpeter) [reply web/src/features/monitors/components/MonitorForm.tsx:498]: Fixed in 11c5a6981: MonitorForm now keeps filters across view changes, matching WidgetForm — `resolveViewChangePatch` only patches the metric, and its return type no longer includes `filters` so the clear cannot be reintroduced. `MetricsFilterBuilder` banners now-invalid rows; submit and preview already strip them.
